cpp/tests/c_api/ecg_test.c

```diff
@@ -0,0 +1,205 @@
+  TEST_ALWAYS_ASSERT(ret_code == CUGRAPH_SUCCESS, "cugraph_ecg failed.");
+
+  if (test_ret_value == 0) {
+    cugraph_type_erased_device_array_view_t* vertices;
+    cugraph_type_erased_device_array_view_t* clusters;
+
+    vertices          = cugraph_hierarchical_clustering_result_get_vertices(result);
+    clusters          = cugraph_hierarchical_clustering_result_get_clusters(result);
+    double modularity = cugraph_hierarchical_clustering_result_get_modularity(result);
+
+    vertex_t h_vertices[num_vertices];
+    edge_t h_clusters[num_vertices];
+
+    ret_code = cugraph_type_erased_device_array_view_copy_to_host(
+      handle, (byte_t*)h_vertices, vertices, &ret_error);
+    TEST_ASSERT(test_ret_value, ret_code == CUGRAPH_SUCCESS, "copy_to_host failed.");
+
+    ret_code = cugraph_type_erased_device_array_view_copy_to_host(
+      handle, (byte_t*)h_clusters, clusters, &ret_error);
+    TEST_ASSERT(test_ret_value, ret_code == CUGRAPH_SUCCESS, "copy_to_host failed.");
+
+    // Louvain and permute_range are both tested, here we only make
+    // sure that function calls succeed as expected.
+
+    cugraph_hierarchical_clustering_result_free(result);
+  }
+
+  cugraph_sg_graph_free(graph);
+  cugraph_free_resource_handle(handle);
+  cugraph_error_free(ret_error);
+
+  return test_ret_value;
+}
+
+int test_ecg()
+{
+  size_t num_edges     = 16;
+  size_t num_vertices  = 6;
+  size_t max_level     = 10;
+  weight_t threshold   = 1e-7;
+  weight_t resolution  = 1.0;
+  weight_t min_weight  = 0.001;
+  size_t ensemble_size = 10;
+
+  vertex_t h_src[] = {0, 1, 1, 2, 2, 2, 3, 4, 1, 3, 4, 0, 1, 3, 5, 5};
+  vertex_t h_dst[] = {1, 3, 4, 0, 1, 3, 5, 5, 0, 1, 1, 2, 2, 2, 3, 4};
+  weight_t h_wgt[] = {
+    0.1f, 2.1f, 1.1f, 5.1f, 3.1f, 4.1f, 7.2f, 3.2f, 0.1f, 2.1f, 1.1f, 5.1f, 3.1f, 4.1f, 7.2f, 3.2f};
+  vertex_t h_result[]          = {0, 0, 0, 1, 1, 1};
+  weight_t expected_modularity = 0.215969;
+
+  // Louvain wants store_transposed = FALSE
+  return generic_ecg_test(h_src,
+                          h_dst,
+                          h_wgt,
+                          h_result,
+                          expected_modularity,
+                          num_vertices,
+                          num_edges,
+                          min_weight,
+                          ensemble_size,
+                          max_level,
+                          threshold,
+                          resolution,
+                          FALSE);
+}
+
+int test_ecg_no_weight()
+{
+  size_t num_edges     = 16;
+  size_t num_vertices  = 6;
+  size_t max_level     = 10;
+  weight_t threshold   = 1e-7;
+  weight_t resolution  = 1.0;
+  weight_t min_weight  = 0.001;
+  size_t ensemble_size = 10;
+
+  vertex_t h_src[]             = {0, 1, 1, 2, 2, 2, 3, 4, 1, 3, 4, 0, 1, 3, 5, 5};
+  vertex_t h_dst[]             = {1, 3, 4, 0, 1, 3, 5, 5, 0, 1, 1, 2, 2, 2, 3, 4};
+  vertex_t h_result[]          = {1, 1, 1, 2, 0, 0};
+  weight_t expected_modularity = 0.0859375;
```

Comment:
Since `h_result` and `expected_modularity` are not used or tested against, I am not sure it is necessary to have them.

Reply:
Removed.